Survey staff view of responses

jchiam opened the pull request:
Implemented staff view of response page with unsubmit feature #1020

kxmbrian commented on app/views/course/survey/responses/_response.json.jbuilder:
```diff
@@ -5,9 +5,11 @@ json.survey do
   end
 end
+
+json.is_creator current_user.id == @survey.creator_id
```
Any staff member or coursemology admin should be able to unsubmit for a student. Can do
`json.isStaff = can?(:manage, @survey)` instead.

The abilities are defined in `Course::SurveysAbilityComponent`. Also, currently I'm using snake_case for model attributes since it's cleaner to use `json.(instance, :attr_name)`, and camelCase for derived attributes since it's the JS convention -- save some renaming in the frontend for those.

jczjchiam replied:
Fixed

kxmbrian commented on client/app/bundles/course/survey/pages/ResponseShow/index.jsx:
```diff
@@ -119,8 +127,35 @@ class ResponseShow extends React.Component {
+          <TableRow selectable={false}>
+            <TableHeaderColumn>Submitted At</TableHeaderColumn>
+            <TableRowColumn>
+              <FormattedDate
```
Could you use `moment` instead? We want to move away from `FormattedDate` because it's clunky to use.

jczjchiam replied:
Changed `FormattedDate` in favour of `moment`

weiqingtoh commented on app/controllers/course/survey/responses_controller.rb:
```diff
@@ -28,7 +28,12 @@ def edit
+    if params[:response][:submit]
+      @response.submit
+    elsif params[:response][:unsubmit]
+      @response.unsubmit
```
Small concern: since `unsubmit` is more of a single action - however, it is actually possible for a user to pass in updated response params, which seems as though it would get updated here as well. Even if the react components don't pass these params into the controller, it might still be cleaner to ensure the controller doesn't do this.

allenwq replied:
we can filter out the params in strong params ( based on ability ), so for students this param should be filtered out.

weiqingtoh commented on client/app/bundles/course/survey/pages/ResponseShow/ResponseForm.jsx:
```diff
@@ -113,35 +123,62 @@ class ResponseForm extends React.Component {
-          disabled={!!response.submitted_at}
-        />
+        {isCreator ? null : this.renderSaveButton()}
+        {isCreator ? this.renderUnsubmitButton() : this.renderSubmitButton()}
```
Should this be `isSubmitted` or something?

jchiam commented:
@kxmbrian @allenwq please re-review

@weiqingtoh Not sure if I'm doing the params filtering right.
https://github.com/coursemology-collab/coursemology2/blob/2d8ae73534dfe84584f498bd3cbf91daaaee6dad/app/controllers/course/survey/responses_controller.rb#L31-L40

allenwq commented on app/controllers/course/survey/responses_controller.rb:
```diff
@@ -28,11 +28,22 @@ def edit
-      render_response_json
-    else
-      render json: { errors: @response.errors }, status: :bad_request
+    if params[:response][:unsubmit]
```
Must check the permission of `unsubmit`, otherwise student can actually send requests to unsubmit their own surveys.

This can be done through strong params, something like:
```
permitted_params = []
if can?(:update, @survey)
   permitted_params << :unsubmit
end
params.require(:response).permit(permmited_params)
```

allenwq replied:
😄  @jchiam sorry didn't notice your comment 

>Not sure if I'm doing the params filtering right.
I have an example of param filtering above

jchiam replied:
Update to do the permissions check

weiqingtoh replied:
Can I suggest that since `unsubmit` is rather different from `update`, we just implement it in a separate method? Otherwise we are trying to do too many things in this controller method. Understand some rework is required for the react side. Thoughts? @kxmbrian @allenwq

jchiam replied:
@weiqingtoh Correct me if I'm wrong, but both submit and unsubmit creates patch request which is handled by `update` in the controller. So at best you mean I should refactor `update` such that it calls a method for submit and another for unsubmit?

allenwq commented on client/app/bundles/course/survey/store.js:
```diff
@@ -5,7 +5,7 @@ import rootReducer from './reducers';
   const initialStates = surveys;
   const storeCreator = (process.env.NODE_ENV === 'development') ?
-    compose(applyMiddleware(thunkMiddleware, require('redux-logger').logger))(createStore) : // eslint-disable-line global-require
+    compose(applyMiddleware(thunkMiddleware, require('redux-logger').logger))(createStore) : // eslint-disable-line global-require, max-len
```
is `max-len` here duplicated with the one in eslint config ?

jchiam replied:
what do you mean by that? this line triggered the max-len warning because of the disable line comment. i added `max-len` as well to stop the warning for this line only

kxmbrian replied:
if it is neater, maybe can do `eslint-disable-next-line`?

jchiam replied:
fixed this

kxmbrian commented on app/views/course/survey/responses/_response.json.jbuilder:
```diff
@@ -5,9 +5,11 @@ json.survey do
 
 json.response do
-  json.(response, :id, :submitted_at)
+  json.(response, :id, :submitted_at, :creator)
```
This will serialize the whole `User` instance and send it over, including fields like `authentication_token`. Better to extract only the fields that we need.

Also good to update the documentation at `client/app/api/course/Survey/Responses.js`.

allenwq commented on app/controllers/course/survey/responses_controller.rb:
```diff
@@ -37,11 +37,22 @@ def edit
+    if params[:response][:unsubmit]
+      @response.unsubmit if can?(:update, @survey)
+
+      if @response.update_attributes(response_unsubmit_params)
```
@jchiam any reasons why `response_update_params` cannot be used here ?

jchiam replied:
> from @weiqingtoh Small concern: since unsubmit is more of a single action - however, it is actually possible for a user to pass in updated response params, which seems as though it would get updated here as well. Even if the react components don't pass these params into the controller, it might still be cleaner to ensure the controller doesn't do this.

`response_unsubmit_params` will not allow any of the form params to update

kxmbrian commented:
Could you add a confirmation dialogue for the submission? Here's the one for assessments for reference:

![screen shot 2017-04-03 at 5 39 24 pm](https://cloud.githubusercontent.com/assets/6252919/24603751/a8021668-1894-11e7-9737-4c841b2974f0.png)

Also, can disable the form fields for staff view?

kxmbrian commented on client/app/bundles/course/survey/pages/ResponseShow/ResponseForm.jsx:
```diff
@@ -113,35 +124,82 @@ class ResponseForm extends React.Component {
+  shouldDisableSaveButton() {
+    const { pristine, response, isResponseCreator } = this.props;
+
+    if (!!response.submitted_at || !!pristine || !isResponseCreator) {
```
`pristine` is already a bool, don't need to `!!`.

kxmbrian commented on client/app/bundles/course/survey/pages/ResponseShow/index.jsx:
```diff
@@ -38,6 +38,18 @@ const translations = defineMessages({
+  loading: {
+    id: 'course.surveys.ResponseShow.loading',
+    defaultMessage: 'Loading survey questions...',
+  },
```
`loading` is not used anymore. can remove it!

kxmbrian commented on client/app/bundles/course/survey/pages/ResponseShow/index.jsx:
```diff
@@ -117,6 +130,18 @@ class ResponseShow extends React.Component {
+    const { saveSuccess, saveFailure, unsubmitSuccess, unsubmitFailure } = translations;
+    const payload = ResponseShow.formatSurveyResponseData(data);
+    const successMessage = <FormattedMessage {...(data.unsubmit ? unsubmitSuccess : saveSuccess)} />;
+    const failureMessage = <FormattedMessage {...(data.unsubmit ? unsubmitFailure : saveFailure)} />;
```
`handleUpdateResponse` handles both save and submit, hence the `data.submit ? ... : ...` switches. We don't need it for `renderUnsubmitButton` since it should be handling one case only.

Don't think we want to allow staff to update the other fields as well. We can construct the payload directly as 
```
const payload = { response: { unsubmit: true } }
```

kxmbrian commented on app/controllers/course/survey/responses_controller.rb:
```diff
@@ -37,11 +37,22 @@ def edit
+    if params[:response][:unsubmit]
+      @response.unsubmit if can?(:manage, @survey)
+
+      if @response.update_attributes(response_unsubmit_params)
```
Since we don't intend for users to unsubmit and update fields at the same time, we can just save the response:

```
if @response.save
```

kxmbrian commented:
Aesthetics-wise, maybe can make the width of the table say 60% the width of the page to being the keys and values closer to each other and disable `stripedRows` so that it looks less like a header?

@allenwq @weiqingtoh do you think we should break the convention of having the unsubmit button in bright red? Since it is a 'last resort' kind function, I think we should make it less conspicuous.

jchiam commented:
@kxmbrian @allenwq @weiqingtoh  Please re-review

@kxmbrian Can I do the unsubmit confirmation dialog in the next PR instead. This PR is getting fat.

weiqingtoh commented on client/app/bundles/course/survey/pages/ResponseShow/ResponseForm.jsx:
```diff
@@ -114,35 +124,82 @@ class ResponseForm extends React.Component {
-          disabled={!!response.submitted_at}
-        />
+        {this.renderSaveButton()}
+        {this.renderSubmitButton()}
```
Once submitted can the student still view the response?

jchiam replied:
Yes

kxmbrian commented on client/app/bundles/course/survey/pages/ResponseShow/ResponseForm.jsx:
```diff
@@ -11,6 +11,10 @@ const styles = {
   },
+  unsubmitButton: {
+    marginRight: 10,
+    backgroundColor: 'red',
```
Sorry, missed this earlier. Could you `import { red500 } from 'material-ui/styles/colors';` and do `backgroundColor: red500`?

kxmbrian commented:
Other than the lints, LGTM. 

For now, I think it's fine to hide the save button when submitted. The logic will change once we port the "Allow modify/save after submit" feature.

Am okay with doing the confirmation dialog and tests in the separate PRs!

I missed out another thing: could we disable the form in staff view? If you want, can do in next PR also.

jchiam commented:
Latest push:
@weiqingtoh @allenwq 
Spaced out the form buttons
Added new logic to showing the buttons
- If not response creator, save and submit buttons are hidden
- For response creator, save button is hidden if submitted

@kxmbrian 
Fixed lint errors
Coming up in future PRs:
- Disable form fields appropriately
- Confirmation dialog for unsubmit
- Tests

allenwq commented:
@jchiam One of the existing tests failed, can you check ? 

LGTM otherwise.

jchiam commented:
Tests fixed! 😄

weiqingtoh commented:
We need some tests for the backend and controller - `unsubmit`, and to test when `unsubmit` doesn't change existing params (in case of future refactoring). @jchiam you can rest first, if need be I can write the tests.

kxmbrian commented:
LGTM!

allenwq commented:
Thanks for helping @jchiam ! 🎉
